File: packages/sdk-py/src/agenttool/authority.py
"""Agent-held constitutional HTTP mutation and exact-private-read proofs.

Byte-identical to ``api/src/services/identity/authority.ts``.
"""

from __future__ import annotations

import base64
import hashlib
from datetime import datetime, timezone
from typing import Any, Dict, Mapping, TypedDict, Union
from urllib.parse import urlsplit

from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey


IDENTITY_AUTHORITY_DOMAIN = "identity-authority/v1"
IDENTITY_READ_AUTHORITY_DOMAIN = "identity-read-authority/v1"
AUTHORITY_SEQUENCE_HEADER = "X-Agenttool-Authority-Sequence"
AUTHORITY_TIMESTAMP_HEADER = "X-Agenttool-Authority-Timestamp"
AUTHORITY_SIGNATURE_HEADER = "X-Agenttool-Authority-Signature"


def authority_timestamp_now() -> str:
    """Default freshness stamp for a proof. The server window is ±5 minutes."""
    return (
        datetime.now(timezone.utc)
        .isoformat(timespec="milliseconds")
        .replace("+00:00", "Z")
    )


def authority_request_target(url: str) -> str:
    """Return the exact origin-form request target an authority proof covers.

    Byte-identical to ``authorityRequestTarget`` in the API service. The proof
    binds path-and-query, so derive it from the same absolute URL the
    transport will actually fetch.
    """
    parsed = urlsplit(url)
    return f"{parsed.path}?{parsed.query}" if parsed.query else parsed.path


def canonical_identity_authority_bytes(
    *,
    identity_did: str,
    method: str,
    request_target: str,
    body: Union[str, bytes, bytearray],
    sequence: int,
    timestamp: str,
) -> bytes:
    """Return the 32-byte ``identity-authority/v1`` digest to sign.

    ``body`` must be the exact bytes sent as the request entity. Serialize
    JSON once, pass that string here, then send the same string as data.
    """
    if not isinstance(sequence, int) or isinstance(sequence, bool) or sequence < 1:
        raise ValueError("authority sequence must be a positive integer")
    if not request_target.startswith("/") or "#" in request_target:
        raise ValueError(
            "authority request_target must be an absolute path with optional query and no fragment"
        )
    body_bytes = body.encode("utf-8") if isinstance(body, str) else bytes(body)
    body_hash_hex = hashlib.sha256(body_bytes).hexdigest()
    fields = [
        identity_did,
        method.upper(),
        request_target,
        body_hash_hex,
        str(sequence),
        timestamp,
    ]
    joined = IDENTITY_AUTHORITY_DOMAIN.encode("utf-8")
    for field in fields:
        joined += b"\x00" + field.encode("utf-8")
    return hashlib.sha256(joined).digest()


def identity_authority_headers(
    *,
    identity_did: str,
    method: str,
    request_target: str,
    body: Union[str, bytes, bytearray],
    sequence: int,
    timestamp: str,
    signing_key: bytes,
) -> Dict[str, str]:
    """Sign exact request bytes and return the three authority headers."""
    if not isinstance(signing_key, (bytes, bytearray)) or len(signing_key) != 32:
        raise ValueError("signing_key must be a 32-byte ed25519 seed")
    canonical = canonical_identity_authority_bytes(
        identity_did=identity_did,
        method=method,
        request_target=request_target,
        body=body,
        sequence=sequence,
        timestamp=timestamp,
    )
    signature = Ed25519PrivateKey.from_private_bytes(bytes(signing_key)).sign(canonical)
    return {
        AUTHORITY_SEQUENCE_HEADER: str(sequence),
        AUTHORITY_TIMESTAMP_HEADER: timestamp,
        AUTHORITY_SIGNATURE_HEADER: base64.b64encode(signature).decode("ascii"),
    }


class AuthorityBinding(TypedDict, total=False):
    """An agent-rooted identity's consent to one exact HTTP mutation.

    ``IdentityAuthority`` in identity.py is this shape. It lives here so that
    any client — identity, at-rest, memory — can bind a proof without
    importing the identity module.
    """

    #: DID of the identity whose immutable root consents.
    did: str
    #: That identity's immutable root ed25519 seed (32 bytes).
    signing_key: bytes
    #: ``next_sequence`` from ``GET /v1/identities/:id/authority``.
    sequence: int
    #: ISO-8601 UTC instant. Defaults to now; the server window is ±5 minutes.
    timestamp: str


def authority_headers_for_request(
    *,
    method: str,
    url: str,
    body: Union[str, bytes, bytearray],
    authority: Mapping[str, Any],
) -> Dict[str, str]:
    """Return the three proof headers for the exact request about to be sent.

    ``authority`` is an :class:`AuthorityBinding` — ``did``, ``signing_key``,
    ``sequence``, and an optional ``timestamp``.

    ``url`` must be the absolute URL the transport will fetch, query string
    included, and ``body`` the exact entity it will carry (``b""`` for a
    body-less mutation, which is what the server reads there). The proof
    hashes those bytes: serialize once, call this, and transmit the same
    value unchanged — any re-serialization in between invalidates it.
    """
    return identity_authority_headers(
        identity_did=authority["did"],
        method=method,
        request_target=authority_request_target(url),
        body=body,
        sequence=authority["sequence"],
        timestamp=authority.get("timestamp") or authority_timestamp_now(),
        signing_key=authority["signing_key"],
    )


def canonical_identity_read_authority_bytes(
    *,
    identity_did: str,
    request_target: str,
    current_sequence: int,
    timestamp: str,
) -> bytes:
    """Return the digest for one exact, non-consuming private GET proof.

    The HTTP method is always ``GET`` and its request body is always empty.
    ``current_sequence`` may be zero and is bound without being advanced.
    """
    if (
        not isinstance(current_sequence, int)
        or isinstance(current_sequence, bool)
        or current_sequence < 0
        or current_sequence > 9_007_199_254_740_991
    ):
        raise ValueError(
            "read authority current_sequence must be a non-negative safe integer"
        )
    if not request_target.startswith("/") or "#" in request_target:
        raise ValueError(
            "read authority request_target must be an absolute path with optional query and no fragment"
        )
    fields = [
        identity_did,
        "GET",
        request_target,
        hashlib.sha256(b"").hexdigest(),
        str(current_sequence),
        timestamp,
    ]
    joined = IDENTITY_READ_AUTHORITY_DOMAIN.encode("utf-8")
    for field in fields:
        joined += b"\x00" + field.encode("utf-8")
    return hashlib.sha256(joined).digest()


def identity_read_authority_headers(
    *,
    identity_did: str,
    request_target: str,
    current_sequence: int,
    timestamp: str,
    signing_key: bytes,
) -> Dict[str, str]:
    """Sign one exact private GET without consuming the mutation sequence."""
    if not isinstance(signing_key, (bytes, bytearray)) or len(signing_key) != 32:
        raise ValueError("signing_key must be a 32-byte ed25519 seed")
    canonical = canonical_identity_read_authority_bytes(
        identity_did=identity_did,
        request_target=request_target,
        current_sequence=current_sequence,
        timestamp=timestamp,
    )
    signature = Ed25519PrivateKey.from_private_bytes(bytes(signing_key)).sign(canonical)
    return {
        AUTHORITY_SEQUENCE_HEADER: str(current_sequence),
        AUTHORITY_TIMESTAMP_HEADER: timestamp,
        AUTHORITY_SIGNATURE_HEADER: base64.b64encode(signature).decode("ascii"),
    }

File: packages/sdk-ts/src/authority.ts
/** Agent-held constitutional HTTP mutation and exact-private-read proofs.
 * Byte-identical to api/src/services/identity/authority.ts. */

import * as ed from "@noble/ed25519";
import { sha256, sha512 } from "@noble/hashes/sha2.js";

ed.etc.sha512Sync = (...m: Uint8Array[]) => {
  const h = sha512.create();
  for (const msg of m) h.update(msg);
  return h.digest();
};

export const IDENTITY_AUTHORITY_DOMAIN = "identity-authority/v1";
export const IDENTITY_READ_AUTHORITY_DOMAIN = "identity-read-authority/v1";
export const AUTHORITY_HEADERS = Object.freeze({
  sequence: "X-Agenttool-Authority-Sequence",
  timestamp: "X-Agenttool-Authority-Timestamp",
  signature: "X-Agenttool-Authority-Signature",
});

export interface CanonicalIdentityAuthorityOpts {
  identityDid: string;
  method: string;
  /** Origin-form path plus its exact query string, if any. */
  requestTarget: string;
  /** The exact bytes that will be transmitted as the HTTP request entity. */
  body: string | Uint8Array;
  sequence: number;
  timestamp: string;
}

export interface CanonicalIdentityReadAuthorityOpts {
  identityDid: string;
  /** Origin-form GET path plus its exact query string, if any. */
  requestTarget: string;
  /** The identity's current mutation sequence. Reads do not advance it. */
  currentSequence: number;
  timestamp: string;
}

function bytes(value: string | Uint8Array): Uint8Array {
  return typeof value === "string" ? new TextEncoder().encode(value) : value;
}

function hex(value: Uint8Array): string {
  return Array.from(value, (b) => b.toString(16).padStart(2, "0")).join("");
}

function b64(value: Uint8Array): string {
  let binary = "";
  for (const byte of value) binary += String.fromCharCode(byte);
  return globalThis.btoa(binary);
}

/** Default freshness stamp for a proof. The server window is ±5 minutes. */
export function authorityTimestampNow(): string {
  return new Date().toISOString();
}

/** Exact origin-form request target an authority proof covers.
 *  Byte-identical to `authorityRequestTarget` in the API service — the
 *  proof binds path-and-query, so the caller must derive it from the same
 *  absolute URL the transport will actually fetch. */
export function authorityRequestTarget(url: string): string {
  const parsed = new URL(url);
  return `${parsed.pathname}${parsed.search}`;
}

export function canonicalIdentityAuthorityBytes(
  opts: CanonicalIdentityAuthorityOpts,
): Uint8Array {
  if (!Number.isSafeInteger(opts.sequence) || opts.sequence < 1) {
    throw new Error("authority sequence must be a positive safe integer");
  }
  if (!opts.requestTarget.startsWith("/") || opts.requestTarget.includes("#")) {
    throw new Error(
      "authority requestTarget must be an absolute path with optional query and no fragment",
    );
  }
  const enc = new TextEncoder();
  const fields = [
    enc.encode(opts.identityDid),
    enc.encode(opts.method.toUpperCase()),
    enc.encode(opts.requestTarget),
    enc.encode(hex(sha256(bytes(opts.body)))),
    enc.encode(String(opts.sequence)),
    enc.encode(opts.timestamp),
  ];
  const parts: Uint8Array[] = [enc.encode(IDENTITY_AUTHORITY_DOMAIN)];
  for (const field of fields) parts.push(new Uint8Array([0]), field);
  const length = parts.reduce((sum, part) => sum + part.length, 0);
  const joined = new Uint8Array(length);
  let offset = 0;
  for (const part of parts) {
    joined.set(part, offset);
    offset += part.length;
  }
  return sha256(joined);
}

export function identityAuthorityHeaders(
  opts: CanonicalIdentityAuthorityOpts & { signingKey: Uint8Array },
): Record<string, string> {
  if (opts.signingKey.length !== 32) {
    throw new Error("signingKey must be a 32-byte ed25519 seed");
  }
  const signature = b64(
    ed.sign(canonicalIdentityAuthorityBytes(opts), opts.signingKey),
  );
  return {
    [AUTHORITY_HEADERS.sequence]: String(opts.sequence),
    [AUTHORITY_HEADERS.timestamp]: opts.timestamp,
    [AUTHORITY_HEADERS.signature]: signature,
  };
}

/**
 * An agent-rooted identity's consent to one exact HTTP mutation.
 *
 * `IdentityAuthority` in identity.ts is this shape, and is accepted here
 * structurally. It lives in this module so that any client — identity,
 * at-rest, memory — can bind a proof without importing the identity module.
 */
export interface AuthorityBinding {
  /** DID of the identity whose immutable root consents. */
  did: string;
  /** That identity's immutable root ed25519 seed (32 bytes). */
  signing_key: Uint8Array;
  /** `next_sequence` from `GET /v1/identities/:id/authority`. */
  sequence: number;
  /** ISO-8601 UTC instant. Defaults to now; the server window is ±5 minutes. */
  timestamp?: string;
}

/**
 * The three proof headers for the exact request a client is about to send.
 *
 * Pass the absolute URL the transport will fetch and the exact entity it
 * will carry — the empty string for a body-less mutation, which is what the
 * server reads there. The proof hashes those bytes, so a client must
 * serialize once, call this, and transmit the same value unchanged; any
 * re-serialization in between invalidates the signature.
 */
export function authorityHeadersForRequest(opts: {
  method: string;
  /** Absolute URL the transport will fetch, query string included. */
  url: string;
  /** Exact entity bytes that will be transmitted. */
  body: string | Uint8Array;
  authority: AuthorityBinding;
}): Record<string, string> {
  return identityAuthorityHeaders({
    identityDid: opts.authority.did,
    method: opts.method,
    requestTarget: authorityRequestTarget(opts.url),
    body: opts.body,
    sequence: opts.authority.sequence,
    timestamp: opts.authority.timestamp ?? authorityTimestampNow(),
    signingKey: opts.authority.signing_key,
  });
}

/**
 * Return the exact digest for an identity-read-authority/v1 proof.
 *
 * Private reads are deliberately GET-only with an empty request body. The
 * current mutation sequence (including zero) is bound but never consumed.
 */
export function canonicalIdentityReadAuthorityBytes(
  opts: CanonicalIdentityReadAuthorityOpts,
): Uint8Array {
  if (
    !Number.isSafeInteger(opts.currentSequence) ||
    opts.currentSequence < 0
  ) {
    throw new Error(
      "read authority currentSequence must be a non-negative safe integer",
    );
  }
  if (!opts.requestTarget.startsWith("/") || opts.requestTarget.includes("#")) {
    throw new Error(
      "read authority requestTarget must be an absolute path with optional query and no fragment",
    );
  }
  const enc = new TextEncoder();
  const fields = [
    enc.encode(opts.identityDid),
    enc.encode("GET"),
    enc.encode(opts.requestTarget),
    enc.encode(hex(sha256(new Uint8Array()))),
    enc.encode(String(opts.currentSequence)),
    enc.encode(opts.timestamp),
  ];
  const parts: Uint8Array[] = [enc.encode(IDENTITY_READ_AUTHORITY_DOMAIN)];
  for (const field of fields) parts.push(new Uint8Array([0]), field);
  const length = parts.reduce((sum, part) => sum + part.length, 0);
  const joined = new Uint8Array(length);
  let offset = 0;
  for (const part of parts) {
    joined.set(part, offset);
    offset += part.length;
  }
  return sha256(joined);
}

/** Sign one exact private GET without consuming the mutation sequence. */
export function identityReadAuthorityHeaders(
  opts: CanonicalIdentityReadAuthorityOpts & { signingKey: Uint8Array },
): Record<string, string> {
  if (opts.signingKey.length !== 32) {
    throw new Error("signingKey must be a 32-byte ed25519 seed");
  }
  const signature = b64(
    ed.sign(canonicalIdentityReadAuthorityBytes(opts), opts.signingKey),
  );
  return {
    [AUTHORITY_HEADERS.sequence]: String(opts.currentSequence),
    [AUTHORITY_HEADERS.timestamp]: opts.timestamp,
    [AUTHORITY_HEADERS.signature]: signature,
  };
}
